libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/building/BuildingProcessor.kt

```diff
@@ -48,18 +52,33 @@ internal object BuildingProcessor {
     fun queryBuildingOnWaypoint(
         action: BuildingAction.QueryBuildingOnWaypoint
     ): BuildingResult.GetDestination {
-        val waypoints = action.progress.navigationRoute.internalWaypoints()
         val waypointIndex = action.progress.currentLegProgress?.legIndex!! + 1
-        val waypoint = waypoints.filter { it.isLegWaypoint() }.getOrNull(waypointIndex)
-        return BuildingResult.GetDestination(waypoint?.target ?: waypoint?.location)
+        val buildingLocation = getBuildingLocation(waypointIndex, action.progress)
+        return BuildingResult.GetDestination(buildingLocation)
     }
 
     fun queryBuildingOnFinalDestination(
         action: BuildingAction.QueryBuildingOnFinalDestination
     ): BuildingResult.GetDestination {
-        val lastWaypoint = action.progress.navigationRoute
-            .internalWaypoints()
-            .lastOrNull()
-        return BuildingResult.GetDestination(lastWaypoint?.target ?: lastWaypoint?.location)
+        val waypointIndex = action.progress.navigationRoute.internalWaypoints()
+            .indexOfLast { it.isLegWaypoint() }
+        val buildingLocation = getBuildingLocation(waypointIndex, action.progress)
+        return BuildingResult.GetDestination(buildingLocation)
+    }
+
+    private fun getBuildingLocation(legWaypointIndex: Int, progress: RouteProgress): Point? {
+        val waypoints = progress.navigationRoute.internalWaypoints()
+        val legWaypoints = waypoints.filter { it.isLegWaypoint() }
+        val waypoint = legWaypoints.getOrNull(legWaypointIndex)
+        if (waypoint?.isRequestedWaypoint() != true) {
+            return waypoint?.location
+        }
+        if (waypoint.target != null) {
+            return waypoint.target
+        }
+        val prevEVWaypointsCount = legWaypoints.take(legWaypointIndex + 1)
+            .count { it.isEVChargingWaypoint() }
```

Comment:
nit
it was hard to get the idea at the first glance. Up to you 
```suggestion
        val coordinateIdx = waypoints.map { it.isRequestedWaypoints }.indexOf(waypoints.map {it.isLegWaypoint }[legWaypointIndex]
        return progress.navigationRoute.routeOptions.coordinatesList()
            .getOrNull(coordinateIdx)
```